### DISCUSSION TITLE
Performance of array traversal

### DISCUSSION THREAD

**Discussion by mihaigalos** [vlang/v#13494] [Ideas]
Hi,
I've come across something interesting: performace of the `quicksort` algorithm is at least an order of magnitude faster in `golang` than in `vlang`.

<details>
<summary>vlang - click to expand</summary>

```vlang
import rand
import time

const (
	len = 2000000 // how many random numbers to generate
	max = 999 // max of the generated numbers
)

fn main() {
	mut arr := generate_data(len, max)
	println('length of random array is $arr.len')
	println('is sorted: ${is_sorted<int>(arr)}')
	println('--- Starting ---')

	start := time.now()
	quicksort<int>(mut arr)
	end := time.now()

	println('--- Done ---')
	println('is sorted: ${is_sorted<int>(arr)}')

	print((end - start) / time.second)
	println('s')
}

fn generate_data(length int, maximum int) []int {
	mut arr := []int{cap: length}
	for _ in 0 .. length {
		arr << rand.intn(maximum)
	}

	return arr
}

fn quicksort<T>(mut a []T) {
	if a.len < 2 {
		return
	}

	mut left, mut right := 0, a.len - 1
	mut pivot := rand.intn(max) % a.len
	a[pivot], a[right] = a[right], a[pivot]

	for i, _ in a {
		if a[i] < a[right] {
			a[left], a[i] = a[i], a[left]
			left++
		}
	}

	a[left], a[right] = a[right], a[left]

	quicksort(mut a[..left])
	quicksort(mut a[left + 1..])
}

fn is_sorted<T>(arr []T) bool {
	for i in 0 .. arr.len - 1 {
		if arr[i] > arr[i + 1] {
			return false
		}
	}
	return true
}
```
</details>

<details>
<summary>golang - click to expand</summary>

```golang
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func main() {

	slice := generate_data(2000000)
	fmt.Println("--- Starting ---")
	start := time.Now()
	a := quicksort(slice)
	end := time.Now()
	fmt.Println("--- Done ---")
	fmt.Println("is sorted:", is_sorted(a))
	fmt.Print(end.Sub(start))

}

func generate_data(size int) []int {

	slice := make([]int, size, size)
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < size; i++ {
		slice[i] = rand.Intn(999)
	}
	return slice
}

func quicksort(a []int) []int {
	if len(a) < 2 {
		return a
	}

	left, right := 0, len(a)-1
	pivot := rand.Int() % len(a)
	a[pivot], a[right] = a[right], a[pivot]

	for i, _ := range a {
		if a[i] < a[right] {
			a[left], a[i] = a[i], a[left]
			left++
		}
	}

	a[left], a[right] = a[right], a[left]

	quicksort(a[:left])
	quicksort(a[left+1:])

	return a
}

func is_sorted(a []int) bool {
	for i, _ := range a[0 : len(a)-1] {
		if a[i] > a[i+1] {
			return false
		}
	}
	return true
}
```
</details>

<details>
<summary>Version and HW info</summary>

```bash
$ v version
V 0.2.4 89e9f8a

$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.3 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.3 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

$ lscpu
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   39 bits physical, 48 bits virtual
CPU(s):                          4
On-line CPU(s) list:             0-3
Thread(s) per core:              2
Core(s) per socket:              2
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           60
Model name:                      Intel(R) Core(TM) i5-4570T CPU @ 2.90GHz
Stepping:                        3
CPU MHz:                         1935.031
CPU max MHz:                     2900.0000
CPU min MHz:                     800.0000
BogoMIPS:                        5786.97
Virtualization:                  VT-x
L1d cache:                       64 KiB
L1i cache:                       64 KiB
L2 cache:                        512 KiB
L3 cache:                        4 MiB
NUMA node0 CPU(s):               0-3
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Full generic retpoline, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling
Vulnerability Srbds:             Mitigation; Microcode
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs 
                                 bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt ts
                                 c_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1
                                  avx2 smep bmi2 erms invpcid xsaveopt dtherm arat pln pts md_clear flush_l1d
```
</details>

Since the syntax is almost identical, one can have them side-by-side for comparison. Here's what I got trying to sort the 2M elements:
```
vlang: 19 seconds 
golang: 1.25 seconds
```

I went ahead and generated the corresponding C file with `v quick_sort.v -o quick_sort.c` and saw this:
```cpp
VV_LOCAL_SYMBOL voidptr array_get(array a, int i) {
	..
	return ((byte*)(a.data)) + i * a.element_size;
}
```
I'm wondering if this is the best approach for getters and setters. Of course, this works fine for random access in the slice.

However, I suspect it not best suited for contiguous blocks of memory in the slice.
With i.e. `C++`, in contiguous memory containers such as a `std::vector`, the next element is already prefetched and present in the CPU cache.

If indirection is performed (like here, via the `a[i * a.element_size]`), my suspicion is that the prefetching does not work and the cache is invalidated.

Potential suggested improvements:
* Separate, inline get/set pair for contiguous data to avoid the function call and hence indirection to the slice element.
* Use pass-by-reference as parameters to avoid array copy when calling the get/set.

Thoughts?

**Comment by JalonSolov** [comment]
Still a bit slower than Go, but a simple change mostly fixes it...
```
$ v run qstest.v
length of random array is 2000000
is sorted: false
--- Starting ---
--- Done ---
is sorted: true
3s
$
```
The change is to add the `[direct_array_access]` attribute to the quicksort routine, as in
```v
[direct_array_access]
fn quicksort<T>(mut a []T) {
```

**Reply by mihaigalos** [reply]
Fastest I got was `6s`, which is `3x` improvement from the original. Still about `5x` slower than `go`.

**Comment by ghost** [comment]
with `-prod`?

**Reply by mihaigalos** [reply]
Seems that `-prod` switch and `direct_array_access` tag in the file did it!
It's now down to 1s:
```bash
$ v -prod run quick_sort.v

Use `v run` without optimization, or build an optimized binary with -prod first, then run it separately.
length of random array is 2000000
is sorted: false
--- Starting ---
--- Done ---
is sorted: true
1s
```

Thank you all! This is awesome!

**Comment by ghost** [comment]
Most of the complaints about performance are from people who don't know about `-prod`.
I think this should be rewrite.
https://github.com/vlang/v/blob/master/doc/docs.md#performance-tuning

C code generated without `-prod` is insanely slow.
If you want to compare it with other languages, make sure to add `-prod`.

In order to avoid unnecessary bad impressions, can you fix it to a good sentence and PR it? @JalonSolov

**Reply by mihaigalos** [reply]
What is the reason for generating non-optimized output by default?
Why is the `-prod` not implicit and `-debug` or similar as a compiler flag?

To be clear: these are probably very legitimate design decisions, but as cool as `vlang` is, I find it doesn't just work out-of-the-box in this regard.

**Reply by ghost** [reply]
I don't know the details myself, but one is this.
https://github.com/vlang/v/blob/245b70cabcadfb7216206d4a62cc4c1509be9c41/cmd/v/help/build.txt#L74

The rest of the optimization will take some time.
Default is development mode.
`-prod` is used to deploy the final stage product.

**Reply by mihaigalos** [reply]
True, but the compilation is not a lot longer than `C`.
I can understand this is favoring the developers of `vlang` but seems to not be putting users first.

**Reply by medvednikov** [reply]
> What is the reason for generating non-optimized output by default?

Because it's much faster to compile. And that's how it works with all compilers.

**Reply by medvednikov** [reply]
Well, Go is probably the only exception. Their machine code generator is very fast, and they do optimizations by default.

**Reply by ghost** [reply]
This is better for me.
It takes a lot of time and effort to have to kill all the warnings every time I'm developing.
You only need to do that in the final stage.
Also, the compilation speed should be as fast as possible during development.
This speed is something that Alex is particularly concerned about.
Besides, it's out of the question to compare it with the compile speed of C. It's just too stupidly slow.

**Reply by mihaigalos** [reply]
`vlang` seems to however have an opinion on a lot of stuff, optimization isn't among them.

**Reply by medvednikov** [reply]
what do you mean?

**Reply by mihaigalos** [reply]
I think we can stop the discussion here, you've explained some of the core values which I respect.
Thank you for an awesome language and fast reactive time. Perhaps I'll even contribute some PRs in the future, will definitely continue using it!

**Reply by ylluminate** [reply]
@mihaigalos thanks for sharing your initial findings and seeking to improve the situation!

I would add one more thing here since I don't feel as though [your concern/remark](https://github.com/vlang/v/discussions/13494#discussioncomment-2195933) regarding V "not putting users first" may be sufficiently addressed. First I guess one must ask, who are the "users" of a programming language? One could handily argue that the users are the actual developers. I think this is logical and that the experience should be tuned so as to give the best experience with the most options and flexibility to the user (eg, developer).  I believe wholeheartedly this is what Alex and all of the V team has endeavored to do to date.

Ultimately everything deployed to the public for users' of the programs developed with V (developers; the users of V) will be built with `-prod`, but up until that deployment time it is critical to keep the V users' (developers) workflow clean, smooth and rapid. This, I believe, is the spirit of V. 

Coming from a largely CS (C, C++, etc.) + Ruby background, I will say that I haven't felt that a language has had such positive user-centric goals since the inception and development of Ruby.  I feel a real magnetic pull to V due to this user(developer)-centricity that can lead to great programs for the programs' end users.

One final thought: Performance/speed have not been the perceptually highest priority to date (even though there has been some great work in this regard as folks have time since many of us do perceive performance to be important as witnessed by the user experience with rapid compilation and remarks on the homepage re: C comparability) due to the implementation phase that the language is in. You can expect to see much more in the way of optimization and improved documentation in the coming months as V reaches past 0.3.

You are most warmly welcomed to offer PRs now and always! The one rule of the V community can be summarized as "be nice" (simple and clean like the language) and hopefully we can all keep that in mind as we work together.
